### DISCUSSION TITLE
Danger-swift requers libDanger to be in system directories

### DISCUSSION THREAD

**Discussion by msnazarow** [jdx/mise#3540] [Troubleshooting and bug reports]
**Describe the bug**
when danger-swift is installed via Makefile(also via brew, cause brew runs make install PREFIX=/opt/homebrew/Cellar/danger-swift/3.20.2) it copies libDanger into prefix directories(or system if prefix isn't used) 
And when it is installed via mise, it copies only danger-swift, and it's not enough

**To Reproduce**
```
mise settings experimental=true
mise install danger-swift
git clone https://github.com/msnazarow/DangerMiseCheck
cd DangerMiseCheck
danger-swift local 
```
**Expected behavior**
```
Danger: ⅹ Failing the build, there is 1 fail.
## Failures
Test
```
**Actual behavior**
```
ERROR: Could not find a libDanger to link against at any of: [".build/debug", ".build/x86_64-unknown-linux/debug", ".build/release", "/usr/local/lib/danger", "/opt/homebrew/lib/danger"]
```
**`mise doctor` output**
```
version: 2024.12.3 macos-arm64 (0e849fc 2024-12-08)
activated: yes
shims_on_path: no

build_info:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Sun, 8 Dec 2024 16:17:41 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (x86_64-apple-darwin23.0)

dirs:
  cache: ~/Library/Caches/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml

backends:
  aqua
  asdf
  cargo
  core
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox
  
plugins:

toolset:
  asdf:klundberg/asdf-swiftlint@latest  (missing)
  spm:danger/swift@3.20.2

env_vars:
  MISE_SHELL=zsh

settings:
  experimental  true ~/.config/mise/config.toml

No problems found
```
**Additional context**
Add any other context about the problem here. Consider running mise with `--debug` or `--trace` for extra debug info.

**Comment by syhol** [comment]
Steps to reproduce:
```console
❯ git clone https://github.com/msnazarow/DangerMiseCheck
❯ cd DangerMiseCheck
❯ git checkout master
❯ git checkout danger-check
❯ mise use danger-swift npm:danger
❯ danger-swift local
```
I think it's unlikely that mise will copy anything into system prefix directories for you. But the fact its looking in .build/release is a good sign. I wonder if we can copy that into the install directory. I'm not sure what that path is relative to.

The spm backend is still experimental so this sort of problem is expected.

Also @msnazarow your screenshot upload failed, you might want to try again.

**Comment by msnazarow** [comment]
As i did [here](https://github.com/msnazarow/asdf-danger-swift) you can path PREFIX to [Makefile](https://github.com/danger/swift/blob/master/Makefile) like 
make install PREFIX="${HOME}/.local/share/mise/installs/danger-swift/"

I added ~/.local/share/mise/installs/danger-swift/\(version)/lib this path for search in [danger-swift](https://github.com/danger/swift/pull/633/files)

So first of all, there should be libs somewhere, and that directory should be hardcoded in danger sources

**Comment by msnazarow** [comment]
Screenshot contained lib directory with libDanger and other support swiftmodule files

**Comment by syhol** [comment]
I see @msnazarow has added the lib path `NSHomeDirectory() + "/.local/share/mise/installs/danger-swift/\(version)/lib/danger"` but I don't see that lib path in the installs directory, even after https://github.com/jdx/mise/pull/3441 has been merged. Is there a step missing somewhere?

**Comment by msnazarow** [comment]
@jdx mise run `swift build --configuration release --product danger-swift`, but it builds only danger-swift, not libDanger.dylib

As i mensioned before, you should reference Makefile and [install script](https://github.com/danger/swift/blob/808ad8403c8aa3091ffffebd9db03495a68a9753/Scripts/install.sh#L18)

**Comment by jdx** [comment]
that would only be a solution for this one tool, the backend should support any spm package

**Comment by msnazarow** [comment]
I think you can just call swift build without --product specification

**Comment by Mostafa-Ashraf97** [comment]
Any update regarding this issue? I am still encountering the `missing libDanger` error.

**Comment by Marukome0743** [comment]
This should now be fixed on `main` by #11286.

@jdx's concern above — that a `danger-swift`-only fix wouldn't be right and the
backend should work for any SPM package — is handled by making the fix generic:

- The `spm` backend gained a general `install_command` option. When set, mise runs
  that command from the checked-out package directory instead of discovering
  executable products and running `swift build --product`. It runs with
  `install_env`, the Swift dependency's `PATH`, and both `PREFIX` and
  `MISE_TOOL_INSTALL_PATH` set to the tool's install directory.
  (`install_command` only applies to source installs and can't be combined with
  `filter_bins`. mise still never runs a package's Makefile automatically — it has
  to be configured explicitly.)
- The `danger-swift` registry entry now uses it:
  `{ artifactbundle = false, install_command = "make install PREFIX=\"$MISE_TOOL_INSTALL_PATH\"" }`,
  i.e. Danger Swift's own official install target, so the executable, `libDanger`,
  and the Swift modules all land in the install directory together.

So `danger-swift local` no longer fails with
`Could not find a libDanger to link against ...`:

```sh
mise use danger-swift npm:danger
danger-swift local
```

The e2e test (`e2e/core/test_swift_slow`) asserts that `bin/danger-swift`,
`lib/danger/libDanger.{dylib,so}`, and `Danger.swiftmodule` are all installed, and
that a real `danger-swift local` run reports the failure from the Dangerfile rather
than the `libDanger` error.

Any other SPM package with the same "the Makefile does the real install" shape can
use the option directly:

```toml
[tools]
"spm:owner/repo" = { version = "1.2.3", artifactbundle = false, install_command = "make install PREFIX=\"$MISE_TOOL_INSTALL_PATH\"" }
```

Docs: <https://mise.en.dev/dev-tools/backends/spm.html#install-command>

The change landed after v2026.7.14, so it will ship in the next release.
@msnazarow @Mostafa-Ashraf97 — it'd be great if you could confirm once it's out.

*This comment was generated by an AI coding assistant.*
